File: scripts/audit-data.py
#!/usr/bin/env python3
"""
Data Audit Script - Comprehensive validation of aircraft data
"""

import os
import re
import json
from pathlib import Path
from collections import defaultdict

def main():
    print('🔍 Starting Data Audit...\n')
    
    root_dir = Path(__file__).parent.parent
    
    # 1. Check aircraft data file
    print('📋 Checking aircraft data integrity...')
    data_path = root_dir / 'data' / 'aircraft.ts'
    
    if not data_path.exists():
        print('❌ aircraft.ts not found')
        return False
    
    content = data_path.read_text(encoding='utf-8')
    aircraft_matches = re.findall(r"id:\s*'([^']+)'", content)
    print(f'✅ Found {len(aircraft_matches)} aircraft records\n')
    
    # 2. Check for missing images
    print('🖼️  Checking aircraft images...')
    images_dir = root_dir / 'public' / 'images' / 'aircraft'
    available_images = set()
    
    if images_dir.exists():
        for img_file in images_dir.iterdir():
            if img_file.is_file() and not img_file.name == 'convert.py':
                name_without_ext = img_file.stem
                available_images.add(name_without_ext)
    
    print(f'✅ Found {len(available_images)} image files\n')
    
    # 3. Check category images
    print('📸 Checking category images...')
    categories_dir = root_dir / 'public' / 'images' / 'categories'
    category_images = []
    
    if categories_dir.exists():
        category_images = [f.name for f in categories_dir.iterdir() if f.is_file()]
    
    print(f'✅ Found {len(category_images)} category images')
    print(f'   Images: {", ".join(sorted(category_images[:3]))}... and {len(category_images)-3} more\n')
    
    # 4. Extract aircraft data from TypeScript file
    print('🔬 Validating data structure...')
    
    # Parse aircraft records from TypeScript
    aircraft_pattern = r"\{\s*id:\s*'([^']+)'.*?description:\s*'([^']+)'.*?image:\s*'([^']+)'"
    matches = re.findall(aircraft_pattern, content, re.DOTALL)
    
    missing_images = []
    categories = defaultdict(int)
    engine_types = set()
    countries = set()
    
    # Look for category assignments in data
    category_pattern = r"category:\s*'([^']+)'"
    category_matches = re.findall(category_pattern, content)
    for cat in category_matches:
        categories[cat] += 1
    
    # Extract engine types
    engine_pattern = r"engineType:\s*'([^']+)'"
    engine_matches = re.findall(engine_pattern, content)
    engine_types.update(engine_matches)
    
    # Extract countries
    country_pattern = r"country:\s*'([^']+)'"
    country_matches = re.findall(country_pattern, content)
    countries.update(country_matches)
    
    # Check for missing images
    print('Checking for missing image files...')
    missing_count = 0
    for id_val, desc, image_path in matches:
        if image_path and not image_path.startswith('http'):
            img_name = image_path.split('/')[-1].split('.')[0]
            if img_name not in available_images:
                missing_images.append((id_val, image_path))
                missing_count += 1
    
    if missing_count == 0:
        print('✅ All aircraft images present in filesystem')
    else:
        print(f'⚠️  Found {missing_count} missing images:')
        for aircraft_id, image_path in missing_images[:10]:
            print(f'   - {aircraft_id}: {image_path}')
        if len(missing_images) > 10:
            print(f'   ... and {len(missing_images) - 10} more')
    
    print()
    
    # 5. Validate category references
    print('🏷️  Validating category references...')
    valid_categories = set(categories.keys())
    print(f'✅ Found {len(valid_categories)} categories\n')
    
    # 6. Summary Statistics
    print('📊 Data Summary:')
    print(f'   Total Aircraft: {len(aircraft_matches)}')
    print(f'   Total Categories: {len(categories)}')
    print(f'   Unique Engine Types: {len(engine_types)}')
    print(f'   Countries: {len(countries)}')
    
    print('\n📈 Aircraft by Category:')
    for cat, count in sorted(categories.items(), key=lambda x: -x[1]):
        print(f'   {cat}: {count}')
    
    print('\n✅ Data audit complete!')
    return True

if __name__ == '__main__':
    success = main()
    exit(0 if success else 1)

File: scripts/audit-data.mjs
#!/usr/bin/env node
/**
 * Data Audit Script
 * Validates aircraft data for integrity and missing assets
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const rootDir = path.resolve(__dirname, '..');

console.log('🔍 Starting Data Audit...\n');

// 1. Check aircraft data file
console.log('📋 Checking aircraft data integrity...');
try {
  const dataPath = path.join(rootDir, 'data', 'aircraft.ts');
  const content = fs.readFileSync(dataPath, 'utf-8');
  
  // Extract AIRCRAFT_DATA count
  const exportMatch = content.match(/export const AIRCRAFT_DATA: Aircraft\[\]/);
  if (!exportMatch) {
    console.error('❌ AIRCRAFT_DATA export not found');
    process.exit(1);
  }
  
  // Count aircraft definitions
  const aircraftMatches = content.match(/id: '[^']+'/g) || [];
  console.log(`✅ Found ${aircraftMatches.length} aircraft records\n`);
} catch (error) {
  console.error('❌ Error reading aircraft data:', error.message);
  process.exit(1);
}

// 2. Check for missing images
console.log('🖼️  Checking aircraft images...');
const imagesDir = path.join(rootDir, 'public', 'images', 'aircraft');
const availableImages = new Set(fs.readdirSync(imagesDir).map(f => path.parse(f).name));

console.log(`✅ Found ${availableImages.size} image files\n`);

// 3. Check category images
console.log('📸 Checking category images...');
const categoriesDir = path.join(rootDir, 'public', 'images', 'categories');
const categoryImages = fs.readdirSync(categoriesDir);
console.log(`✅ Found ${categoryImages.length} category images`);
console.log(`   Images: ${categoryImages.join(', ')}\n`);

// 4. Check for data structure issues
console.log('🔬 Validating data structure...');
try {
  // Import the data using file:// URL
  const dataModulePath = path.join(rootDir, 'data', 'aircraft.ts');
  const fileUrl = new URL(`file:///${dataModulePath.replace(/\\/g, '/')}`).href;
  const dataModule = await import(fileUrl);
  const aircraftData = dataModule.AIRCRAFT_DATA || [];
  
  let issues = 0;
  const missingImages = [];
  
  aircraftData.forEach((aircraft, idx) => {
    // Check required fields
    const requiredFields = ['id', 'name', 'category', 'role', 'manufacturer', 'country', 'engineType', 'enginesCount', 'image', 'description', 'specs'];
    
    for (const field of requiredFields) {
      if (!aircraft[field]) {
        console.log(`❌ Aircraft #${idx + 1} (${aircraft.name || 'Unknown'}) missing field: ${field}`);
        issues++;
      }
    }
    
    // Check image exists
    if (aircraft.image && !aircraft.image.startsWith('http')) {
      const imageName = path.parse(aircraft.image).name;
      if (!availableImages.has(imageName)) {
        missingImages.push({
          aircraft: aircraft.name,
          id: aircraft.id,
          image: aircraft.image
        });
      }
    }
    
    // Check specs
    if (aircraft.specs) {
      const specFields = ['wingspan', 'length', 'height', 'payload'];
      for (const field of specFields) {
        if (!aircraft.specs[field]) {
          console.log(`⚠️  Aircraft ${aircraft.name} missing spec: ${field}`);
        }
      }
    }
  });
  
  if (missingImages.length > 0) {
    console.log(`\n⚠️  Missing Image Files (${missingImages.length}):`);
    missingImages.forEach(({ aircraft, id, image }) => {
      console.log(`   - ${aircraft} (${id}): ${image}`);
    });
  } else {
    console.log('✅ All aircraft images present in filesystem');
  }
  
  if (issues === 0) {
    console.log('✅ All aircraft records have required fields\n');
  }
  
  // 5. Validate category references
  console.log('🏷️  Validating category references...');
  const validCategories = new Set(dataModule.CATEGORIES.map(c => c.id));
  let invalidCategoryCount = 0;
  
  aircraftData.forEach(aircraft => {
    if (!validCategories.has(aircraft.category)) {
      console.log(`❌ Aircraft ${aircraft.name} has invalid category: ${aircraft.category}`);
      invalidCategoryCount++;
    }
  });
  
  if (invalidCategoryCount === 0) {
    console.log('✅ All aircraft have valid categories\n');
  }
  
  // 6. Summary Statistics
  console.log('📊 Data Summary:');
  console.log(`   Total Aircraft: ${aircraftData.length}`);
  console.log(`   Total Categories: ${dataModule.CATEGORIES.length}`);
  
  const enginesCount = new Set(aircraftData.map(a => a.engineType));
  console.log(`   Unique Engine Types: ${enginesCount.size}`);
  
  const countries = new Set(aircraftData.map(a => a.country));
  console.log(`   Countries: ${countries.size}`);
  
  const categoryBreakdown = {};
  aircraftData.forEach(a => {
    categoryBreakdown[a.category] = (categoryBreakdown[a.category] || 0) + 1;
  });
  
  console.log('\n📈 Aircraft by Category:');
  Object.entries(categoryBreakdown).sort((a, b) => b[1] - a[1]).forEach(([cat, count]) => {
    console.log(`   ${cat}: ${count}`);
  });
  
  console.log('\n✅ Data audit complete!');
  
} catch (error) {
  console.error('❌ Error during data validation:', error.message);
  process.exit(1);
}
